File: healthcare-portal/src/app/core/services/provider-payout.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface ProviderPayout {
  id: string;
  providerId: string;
  providerName: string;
  periodId: string;
  periodName: string;
  totalAmount: number;
  platformCommission: number;
  providerAmount: number;
  status: string;
  processedAt?: string;
  processedByUserId?: string;
  processedByUserName?: string;
  paymentMethod: string;
  paymentReference?: string;
  notes?: string;
  createdAt: string;
  updatedAt: string;
}

export interface ProcessPayoutDto {
  status: string;
  paymentReference?: string;
  notes?: string;
}

export interface ProviderEarningsDto {
  providerId: string;
  providerName: string;
  totalEarnings: number;
  pendingEarnings: number;
  processedEarnings: number;
  totalConsultations: number;
  averagePerConsultation: number;
  lastPayoutDate?: string;
  nextPayoutDate?: string;
}

export interface PayoutStatisticsDto {
  totalPayouts: number;
  pendingPayouts: number;
  processedPayouts: number;
  totalAmount: number;
  averagePayoutAmount: number;
  totalProviders: number;
  totalPeriods: number;
}

export interface PayoutPeriodDto {
  id: string;
  name: string;
  startDate: string;
  endDate: string;
  status: string;
  totalProviders: number;
  totalAmount: number;
  processedAt?: string;
  createdAt: string;
  updatedAt: string;
}

export interface CreatePayoutPeriodDto {
  name: string;
  startDate: string;
  endDate: string;
}

export interface ApiResponse<T> {
  success: boolean;
  message: string;
  data?: T;
  statusCode: number;
}

@Injectable({
  providedIn: 'root'
})
export class ProviderPayoutService {
  private apiUrl = `${environment.apiUrl}/api/providerpayout`;

  constructor(private http: HttpClient) {}

  // Get payout by ID
  getPayout(id: string): Observable<ApiResponse<ProviderPayout>> {
    return this.http.get<ApiResponse<ProviderPayout>>(`${this.apiUrl}/${id}`);
  }

  // Process payout (Admin only)
  processPayout(id: string, processDto: ProcessPayoutDto): Observable<ApiResponse<ProviderPayout>> {
    return this.http.post<ApiResponse<ProviderPayout>>(`${this.apiUrl}/${id}/process`, processDto);
  }

  // Get payouts by provider
  getPayoutsByProvider(providerId: string): Observable<ApiResponse<ProviderPayout[]>> {
    return this.http.get<ApiResponse<ProviderPayout[]>>(`${this.apiUrl}/provider/${providerId}`);
  }

  // Get payouts by period
  getPayoutsByPeriod(periodId: string): Observable<ApiResponse<ProviderPayout[]>> {
    return this.http.get<ApiResponse<ProviderPayout[]>>(`${this.apiUrl}/period/${periodId}`);
  }

  // Get all payouts with optional filtering
  getAllPayouts(status?: string, page: number = 1, pageSize: number = 50): Observable<ApiResponse<ProviderPayout[]>> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('pageSize', pageSize.toString());
    
    if (status) {
      params = params.set('status', status);
    }
    
    return this.http.get<ApiResponse<ProviderPayout[]>>(`${this.apiUrl}`, { params });
  }

  // Get pending payouts
  getPendingPayouts(): Observable<ApiResponse<ProviderPayout[]>> {
    return this.http.get<ApiResponse<ProviderPayout[]>>(`${this.apiUrl}/pending`);
  }

  // Get payouts by status
  getPayoutsByStatus(status: string): Observable<ApiResponse<ProviderPayout[]>> {
    return this.http.get<ApiResponse<ProviderPayout[]>>(`${this.apiUrl}/status/${status}`);
  }

  // Get provider earnings
  getProviderEarnings(providerId: string): Observable<ApiResponse<ProviderEarningsDto>> {
    return this.http.get<ApiResponse<ProviderEarningsDto>>(`${this.apiUrl}/earnings/${providerId}`);
  }

  // Get payout statistics
  getPayoutStatistics(): Observable<ApiResponse<PayoutStatisticsDto>> {
    return this.http.get<ApiResponse<PayoutStatisticsDto>>(`${this.apiUrl}/statistics`);
  }

  // Generate payouts for period
  generatePayoutsForPeriod(periodId: string): Observable<ApiResponse<boolean>> {
    return this.http.post<ApiResponse<boolean>>(`${this.apiUrl}/period/${periodId}/generate`, {});
  }

  // Process all pending payouts
  processAllPendingPayouts(): Observable<ApiResponse<boolean>> {
    return this.http.post<ApiResponse<boolean>>(`${this.apiUrl}/process-all-pending`, {});
  }

  // Payout Period Management
  createPeriod(createDto: CreatePayoutPeriodDto): Observable<ApiResponse<PayoutPeriodDto>> {
    return this.http.post<ApiResponse<PayoutPeriodDto>>(`${this.apiUrl}/periods`, createDto);
  }

  getPeriod(id: string): Observable<ApiResponse<PayoutPeriodDto>> {
    return this.http.get<ApiResponse<PayoutPeriodDto>>(`${this.apiUrl}/periods/${id}`);
  }

  updatePeriod(id: string, updateDto: CreatePayoutPeriodDto): Observable<ApiResponse<PayoutPeriodDto>> {
    return this.http.put<ApiResponse<PayoutPeriodDto>>(`${this.apiUrl}/periods/${id}`, updateDto);
  }

  getAllPeriods(): Observable<ApiResponse<PayoutPeriodDto[]>> {
    return this.http.get<ApiResponse<PayoutPeriodDto[]>>(`${this.apiUrl}/periods`);
  }

  getActivePeriods(): Observable<ApiResponse<PayoutPeriodDto[]>> {
    return this.http.get<ApiResponse<PayoutPeriodDto[]>>(`${this.apiUrl}/periods/active`);
  }

  deletePeriod(id: string): Observable<ApiResponse<boolean>> {
    return this.http.delete<ApiResponse<boolean>>(`${this.apiUrl}/periods/${id}`);
  }

  processPeriod(id: string): Observable<ApiResponse<boolean>> {
    return this.http.post<ApiResponse<boolean>>(`${this.apiUrl}/periods/${id}/process`, {});
  }
} 

File: backend/SmartTelehealth.Application/Services/ProviderPayoutService.cs
using AutoMapper;
using Microsoft.Extensions.Logging;
using SmartTelehealth.Application.Interfaces;
using SmartTelehealth.Application.DTOs;
using SmartTelehealth.Core.Entities;
using SmartTelehealth.Core.Interfaces;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace SmartTelehealth.Application.Services
{
    public class ProviderPayoutService : IProviderPayoutService
    {
        private readonly IProviderPayoutRepository _providerPayoutRepository;
        private readonly IProviderRepository _providerRepository;
        private readonly IAuditService _auditService;
        private readonly INotificationService _notificationService;
        private readonly IMapper _mapper;
        private readonly ILogger<ProviderPayoutService> _logger;
        public ProviderPayoutService(
            IProviderPayoutRepository providerPayoutRepository,
            IProviderRepository providerRepository,
            IAuditService auditService,
            INotificationService notificationService,
            IMapper mapper,
            ILogger<ProviderPayoutService> logger)
        {
            _providerPayoutRepository = providerPayoutRepository;
            _providerRepository = providerRepository;
            _auditService = auditService;
            _notificationService = notificationService;
            _mapper = mapper;
            _logger = logger;
        }
        public async Task<JsonModel> GetPayoutByIdAsync(Guid id)
        {
            try
            {
                var payout = await _providerPayoutRepository.GetByIdAsync(id);
                if (payout == null)
                {
                    return new JsonModel
                    {
                        data = new object(),
                        Message = "Payout not found",
                        StatusCode = 404
                    };
                }
                var payoutDto = _mapper.Map<ProviderPayoutDto>(payout);
                return new JsonModel
                {
                    data = payoutDto,
                    Message = "Payout retrieved successfully",
                    StatusCode = 200
                };
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error retrieving provider payout");
                return new JsonModel
                {
                    data = new object(),
                    Message = "Failed to retrieve payout",
                    StatusCode = 500
                };
            }
        }
        public async Task<JsonModel> GetPayoutAsync(Guid id) => throw new NotImplementedException();
        public async Task<JsonModel> ProcessPayoutAsync(Guid id, ProcessPayoutDto processDto) => throw new NotImplementedException();
        public async Task<JsonModel> GetPayoutsByProviderAsync(int providerId) => throw new NotImplementedException();
        public async Task<JsonModel> GetPayoutsByPeriodAsync(Guid periodId) => throw new NotImplementedException();
        public async Task<JsonModel> GetAllPayoutsAsync(string status = null, int page = 1, int pageSize = 50)
        {
            try
            {
                IEnumerable<ProviderPayout> payouts;
                if (!string.IsNullOrEmpty(status))
                {
                    payouts = await _providerPayoutRepository.GetByStatusWithPaginationAsync(status, page, pageSize);
                }
                else
                {
                    payouts = await _providerPayoutRepository.GetAllAsync();
                    // Optionally page manually if needed
                    payouts = payouts.Skip((page - 1) * pageSize).Take(pageSize);
                }
                var payoutDtos = _mapper.Map<IEnumerable<ProviderPayoutDto>>(payouts);
                return new JsonModel
                {
                    data = payoutDtos,
                    Message = "Payouts retrieved successfully",
                    StatusCode = 200
                };
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error retrieving payouts");
                return new JsonModel
                {
                    data = new object(),
                    Message = "Failed to retrieve payouts",
                    StatusCode = 500
                };
            }
        }
        public async Task<JsonModel> GetPendingPayoutsAsync()
        {
            try
            {
                var payouts = await _providerPayoutRepository.GetPendingPayoutsAsync();
                var payoutDtos = _mapper.Map<IEnumerable<ProviderPayoutDto>>(payouts);
                return new JsonModel
                {
                    data = payoutDtos,
                    Message = "Pending payouts retrieved successfully",
                    StatusCode = 200
                };
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error retrieving pending payouts");
                return new JsonModel
                {
                    data = new object(),
                    Message = "Failed to retrieve pending payouts",
                    StatusCode = 500
                };
            }
        }
        public async Task<JsonModel> GetPayoutsByStatusAsync(string status)
        {
            try
            {
                var payouts = await _providerPayoutRepository.GetByStatusAsync(status);
                var payoutDtos = _mapper.Map<IEnumerable<ProviderPayoutDto>>(payouts);
                return new JsonModel
                {
                    data = payoutDtos,
                    Message = $"Payouts with status {status} retrieved successfully",
                    StatusCode = 200
                };
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error retrieving payouts by status");
                return new JsonModel
                {
                    data = new object(),
                    Message = "Failed to retrieve payouts by status",
                    StatusCode = 500
                };
            }
        }
        public async Task<JsonModel> GetProviderEarningsAsync(int providerId)
        {
            try
            {
                var provider = await _providerRepository.GetByIdAsync(providerId);
                if (provider == null)
                {
                    return new JsonModel
                    {
                        data = new object(),
                        Message = "Provider not found",
                        StatusCode = 404
                    };
                }
                var totalEarnings = await _providerPayoutRepository.GetTotalPayoutAmountByProviderAsync(providerId);
                var pendingEarnings = await _providerPayoutRepository.GetPendingPayoutAmountByProviderAsync(providerId);
                var payoutCount = await _providerPayoutRepository.GetPayoutCountByProviderAsync(providerId);
                var earnings = new ProviderEarningsDto
                {
                    ProviderId = providerId,
                    ProviderName = $"{provider.FirstName} {provider.LastName}",
                    TotalEarnings = totalEarnings,
                    PendingEarnings = pendingEarnings,
                    ProcessedEarnings = totalEarnings,
                    TotalConsultations = 0, // TODO: Get from consultation service
                    AveragePerConsultation = payoutCount > 0 ? totalEarnings / payoutCount : 0,
                    LastPayoutDate = null, // TODO: Get from payout history
                    NextPayoutDate = null // TODO: Calculate based on payout schedule
                };
                return new JsonModel
                {
                    data = earnings,
                    Message = "Provider earnings retrieved successfully",
                    StatusCode = 200
                };
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error retrieving provider earnings");
                return new JsonModel
                {
                    data = new object(),
                    Message = "Failed to retrieve provider earnings",
                    StatusCode = 500
                };
            }
        }
        public async Task<JsonModel> GetPayoutStatisticsAsync()
        {
            try
            {
                var statistics = await _providerPayoutRepository.GetPayoutStatisticsAsync();
                var statisticsDto = _mapper.Map<PayoutStatisticsDto>(statistics);
                return new JsonModel
                {
                    data = statisticsDto,
                    Message = "Payout statistics retrieved successfully",
                    StatusCode = 200
                };
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error retrieving payout statistics");
                return new JsonModel
                {
                    data = new object(),
                    Message = "Failed to retrieve payout statistics",
                    StatusCode = 500
                };
            }
        }
        public async Task<JsonModel> ProcessAllPendingPayoutsAsync()
        {
            try
            {
                var pendingPayouts = await _providerPayoutRepository.GetPendingPayoutsAsync();
                int processedCount = 0;
                foreach (var payout in pendingPayouts)
                {
                    try
                    {
                        var processDto = new ProcessPayoutDto
                        {
                            Status = "Completed",
                            PaymentReference = $"PAY-{DateTime.UtcNow:yyyyMMdd}-{payout.Id}",
                            Notes = "Automatically processed"
                        };
                        await ProcessPayoutAsync(payout.Id, processDto);
                        processedCount++;
                    }
                    catch (Exception ex)
                    {
                        _logger.LogError(ex, $"Error processing payout {payout.Id}");
                    }
                }
                await _auditService.LogActionAsync("ProviderPayout", "ProcessAll", "System", $"Processed {processedCount} pending payouts");
                return new JsonModel
                {
                    data = true,
                    Message = $"Processed {processedCount} pending payouts",
                    StatusCode = 200
                };
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error processing all pending payouts");
                return new JsonModel
                {
                    data = new object(),
                    Message = "Failed to process pending payouts",
                    StatusCode = 500
                };
            }
        }
        public async Task<JsonModel> GeneratePayoutsForPeriodAsync(Guid periodId) => throw new NotImplementedException();
        public async Task<JsonModel> CreatePeriodAsync(CreatePayoutPeriodDto createDto)
        {
            try
            {
                var period = new PayoutPeriod
                {
                    Name = createDto.Name,
                    StartDate = createDto.StartDate,
                    EndDate = createDto.EndDate,
                    Status = PayoutPeriodStatus.Open,
                    CreatedAt = DateTime.UtcNow,
                    UpdatedAt = DateTime.UtcNow
                };
                var createdPeriod = await _providerPayoutRepository.AddPeriodAsync();
                var periodDto = _mapper.Map<PayoutPeriodDto>(createdPeriod);
                await _auditService.LogActionAsync("PayoutPeriod", "Create", "System", $"Created payout period {createDto.Name}");
                return new JsonModel
                {
                    data = periodDto,
                    Message = "Payout period created successfully",
                    StatusCode = 201
                };
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error creating payout period");
                return new JsonModel
                {
                    data = new object(),
                    Message = "Failed to create payout period",
                    StatusCode = 500
                };
            }
        }
        public async Task<JsonModel> GetAllPeriodsAsync()
        {
            try
            {
                var periods = await _providerPayoutRepository.GetAllPeriodsAsync();
                var periodDtos = _mapper.Map<IEnumerable<PayoutPeriodDto>>(periods);
                return new JsonModel
                {
                    data = periodDtos,
                    Message = "Payout periods retrieved successfully",
                    StatusCode = 200
                };
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error retrieving payout periods");
                return new JsonModel
                {
                    data = new object(),
                    Message = "Failed to retrieve payout periods",
                    StatusCode = 500
                };
            }
        }
    }
} 